File: 2026-01-January-GFG-POTD/January 28 - Count Subset With Target Sum II/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int countSubset(vector<int>& arr, int k) {
        int n = arr.size();
        int mid = n / 2;

        vector<int> left(arr.begin(), arr.begin() + mid);
        vector<int> right(arr.begin() + mid, arr.end());

        unordered_map<int, int> leftSumCount;
        generateSums(left, 0, 0, leftSumCount);

        int count = 0;
        generateAndCount(right, 0, 0, k, leftSumCount, count);

        return count;
    }

private:
    // Generate all subset sums of left half
    void generateSums(vector<int>& arr, int index, int sum,
                      unordered_map<int, int>& mp) {
        if (index == arr.size()) {
            mp[sum]++;
            return;
        }
        generateSums(arr, index + 1, sum, mp);
        generateSums(arr, index + 1, sum + arr[index], mp);
    }

    // Generate sums of right half and count valid pairs
    void generateAndCount(vector<int>& arr, int index, int sum, int k,
                          unordered_map<int, int>& mp, int& count) {
        if (index == arr.size()) {
            count += mp[k - sum];
            return;
        }
        generateAndCount(arr, index + 1, sum, k, mp, count);
        generateAndCount(arr, index + 1, sum + arr[index], k, mp, count);
    }
};

File: 2026-01-January-GFG-POTD/January 28 - Count Subset With Target Sum II/Solution.java
class Solution {
     public int countSubset(int[] arr, int k) {
        int n = arr.length;
        int mid = n / 2;

        int[] left = Arrays.copyOfRange(arr, 0, mid);
        int[] right = Arrays.copyOfRange(arr, mid, n);

        Map<Integer, Integer> leftSumCount = new HashMap<>();
        generateSums(left, 0, 0, leftSumCount);

        int[] count = new int[1];
        generateAndCount(right, 0, 0, k, leftSumCount, count);

        return count[0];
    }

    // Generate all subset sums of left half
    private void generateSums(int[] arr, int index, int sum,
                              Map<Integer, Integer> map) {
        if (index == arr.length) {
            map.put(sum, map.getOrDefault(sum, 0) + 1);
            return;
        }
        generateSums(arr, index + 1, sum, map);
        generateSums(arr, index + 1, sum + arr[index], map);
    }

    // Generate sums of right half and count valid pairs
    private void generateAndCount(int[] arr, int index, int sum, int k,
                                  Map<Integer, Integer> map, int[] count) {
        if (index == arr.length) {
            count[0] += map.getOrDefault(k - sum, 0);
            return;
        }
        generateAndCount(arr, index + 1, sum, k, map, count);
        generateAndCount(arr, index + 1, sum + arr[index], k, map, count);
    }
}
